### DISCUSSION TITLE
merge behavior for unique and deep

### DISCUSSION THREAD

**Discussion by tuxmea** [betadots/hdm#296] [Polls]
There is a misbehaving between Puppet documentation and Puppet results when it comes to having different data types when using unique or deep merge.
See https://github.com/betadots/hdm/pull/283#issuecomment-1929029252

**Comment by oneiros** [comment]
My gut feeling is, that HDM should mirror puppet's behavior as closely as possible, even if that means reimplementing confusing behavior. This means, HDM can be used by developers to analyze exactly what puppet does (or would do).

There is another scenario, though: Maybe HDM is not just meant to analyze and better predict puppet's lookup behavior, but as a kind of linter for hiera yaml files. If we warn in cases where people deviate from best practices, that might be helpful as well.

But I wonder if that is really the case here. At least the behavior of `unique` looks to me as if it might have very legitimate uses. At the very least, it means you can omit cumbersome array syntax in yaml in cases where you only have one value anyway.

**Comment by tuxmea** [comment]
No further feedback. We use the Puppet implementation as is.
